include/mata/nfa.hh

```diff
@@ -493,18 +722,277 @@ inline Nfa uni(const Nfa &lhs, const Nfa &rhs)
     return uni_aut;
 } // uni }}}
 
-void intersection(
-        Nfa*         res,
-        const Nfa&   lhs,
-        const Nfa&   rhs,
-        ProductMap*  prod_map = nullptr);
+/**
+ * Class handling intersection of automata.
+ *
+ * Implements a normal intersection and an intersection preserving epsilon transitions.
+ */
+class Intersection
```

Comment:
I wonder, if it is necessary, to have this functionality encapsulated in separate class.

Reply:
As discussed above, we will move the classes `Intersection` and `Concatenation` to separate `.cc` files. Then, it will not clutter the public interface and can be theoretically refactored into functions.

Nevertheless, I like the current structure better for the following reasons:
- It allows for higher customizability
- The function signatures are much simpler and
- It seems to me more natural to encapsulate similar operations in a class like this than having a multitude of functions without a clear explanation that they belong together.

Reply:
It's a clash of two world.s I'm quite openly anti-OOP, so it's probably just my own preference :-). Your arguments are valid.

Reply:
I fully agree that this depends on our preferences. Do you have an opinion on this, @martinhruska?

Reply:
keep it in classes if you like. Just move it to separate .cc files where it can live its own life.

Reply:
Will do. Since the code already is in classes, I will keep it there.